x-pack/test/plugin_functional/plugins/global_search_test/public/plugin.ts

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { of } from 'rxjs';
+import { map, reduce } from 'rxjs/operators';
+import { Plugin, CoreSetup, CoreStart, AppMountParameters } from 'kibana/public';
+import {
+  GlobalSearchPluginSetup,
+  GlobalSearchPluginStart,
+  NavigableGlobalSearchResult,
+} from '../../../../../plugins/global_search/public';
+import { createResult } from '../common/utils';
+
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface GlobalSearchTestPluginSetup {}
+export interface GlobalSearchTestPluginStart {
+  findAll: (term: string) => Promise<NavigableGlobalSearchResult[]>;
+}
+
+export interface GlobalSearchTestPluginSetupDeps {
+  globalSearch: GlobalSearchPluginSetup;
+}
+export interface GlobalSearchTestPluginStartDeps {
+  globalSearch: GlobalSearchPluginStart;
+}
+
+export class GlobalSearchTestPlugin
+  implements
+    Plugin<
+      GlobalSearchTestPluginSetup,
+      GlobalSearchTestPluginStart,
+      GlobalSearchTestPluginSetupDeps,
+      GlobalSearchTestPluginStartDeps
+    > {
+  public setup(
+    { application, getStartServices }: CoreSetup,
+    { globalSearch }: GlobalSearchTestPluginSetupDeps
+  ) {
+    application.register({
+      id: 'globalSearchTestApp',
+      title: 'GlobalSearch test',
+      mount: (params: AppMountParameters) => {
+        return () => undefined;
+      },
+    });
+
+    globalSearch.registerResultProvider({
+      id: 'gs_test_client',
+      find: (term, options) => {
+        if (term.includes('client')) {
+          return of([
+            createResult({
+              id: 'client1',
+              type: 'test_client_type',
+            }),
+            createResult({
+              id: 'client2',
+              type: 'test_client_type',
+            }),
+          ]);
+        }
+        return of([]);
+      },
+    });
+
+    return {};
+  }
+
+  public start({}: CoreStart, { globalSearch }: GlobalSearchTestPluginStartDeps) {
+    const contract: GlobalSearchTestPluginStart = {
+      findAll: (term) =>
+        globalSearch
+          .find(term, {})
+          .pipe(
+            map((batch) => batch.results),
+            // restrict to test type to avoid failure when real providers are present
+            map((results) => results.filter((r) => r.type.startsWith('test_'))),
+            reduce((memo, results) => [...memo, ...results])
+          )
+          .toPromise(),
+    };
+
+    window.__globalSearchTestApi = contract;
```

Comment:
probably we could put in https://github.com/elastic/kibana/blob/a23a24c8ef3111bb0cced319a53bf104a209c2af/test/plugin_functional/plugins/core_provider_plugin/kibana.json#L5
we can automate it later to declare dependency on all the plugins in the repo

Reply:
Not sure to understand exactly? Should I just add the new `global_search_test` plugin to `core_provider_plugin` optional deps,, or should I move some code to it?

Reply:
> Should I just add the new global_search_test plugin to core_provider_plugin optional deps,

Sorry for being unclear, this one. Then we will use it as a proxy to the plugins API, instead of using a global variable for every separate plugin.

Reply:
As GS is a x-pack plugin, the new test plugin is not in the same folder (`x-pack/test/plugin_functional`), and the test plugin is effectively a x-pack plugin. I don't think I can directly use the `coreProvider` plugin in that case?

Maybe we should have an equivalent in xpack ?

Reply:
> even if only optional, and only for testing

It won't be used unless presents on the file system. I don't think it's a problem and added `licensing` to `CoreProviderPlugin`. YMMV

Reply:
I thought `CoreProviderPlugin` was explicitly tying the exposed contracts. TIL it's an untyped record, so it's fine. Will do the change